### ISSUE TITLE
Docker Support in PCF DEV

### ISSUE THREAD

**Issue by uromahn** [vmware-archive/pcfdev#207]
Not sure if this is an issue with PCF DEV or not, but I nevertheless post this here.

I have enabled Docker support in PCF DEV according to the documentation. When I push a simple docker image directly from Docker Hub, e.g. "cf push registry -o registry:2", the container was successfully created and the service started.

However, when I tag the same Docker image with my local Docker registry which has a self-signed certificate and push the image to it, and use that to do a "cf push registry -o host.pcfdev.io:5000/registry:2", the container was successfully created but after the staging was finished, I am getting an error message "Failed to create container" and the status of the app gets reported as "crashed". It is not clear to me why this would not work since both Docker container are absolutely identical.

What makes this even more problematic is the fact that the provided logging is extremely spares not allowing me to sufficiently debug the problem!

**Comment by sclevine** [comment]
Does your local Docker registry require credentials to pull images? PCF Dev cannot pull Docker images from registries that require authentication.

Also, detailed logs are available inside the VM, in the `/var/vcap/sys/log` directory. You can login to VM using `cf dev ssh`, or you can run `cf dev debug` to extract a snapshot of the logs. The `cf logs` command is intended to provide app logs, not detailed system logs.

**Comment by uromahn** [comment]
Hi Steven,

Thank you for your response. To answer your question:

I am running the default Docker registry image from Docker Hub which does
not require authentication.

I will try to login to the VM and try.to find the logs.

Thank you,
-Uli

On Mar 20, 2017 5:28 PM, "Stephen Levine" <notifications@github.com> wrote:

Does your local Docker registry require credentials to pull images? PCF Dev
cannot pull Docker images from registries that require authentication.

Also, detailed logs are available inside the VM, in the /var/vcap/sys/log
directory. You can login to VM using cf dev ssh, or you can run cf dev debug
to extract a snapshot of the logs. The cf logs command is intended to
provide app logs, not detailed system logs.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/pivotal-cf/pcfdev/issues/207#issuecomment-287939779>,
or mute the thread

**Comment by uromahn** [comment]
I have looked at the logs in garden and noticed the following:

`{"timestamp":"1490020342.806944847","source":"guardian","message":"guardian.api.garden-server.create.failed","log_level":2,"data":{"error":"Get https://host.pcfdev.io:5000/v2: x509: certificate has expired or is not yet valid","request":{"Handle":"6b4f53ed-06a7-4fef-61ca-32f8f4154a66","GraceTime":0,"RootFSPath":"docker://host.pcfdev.io:5000/registry#2","BindMounts":...`

So it complains about the SSL certificate which is really strange since I am able to add this self-signed cert to my other docker daemon and do `docker push` into that.

So, either the `-r` does not work as expected (I did start my cf vm with the `-r host.pcfdev.io:5000` command line) or something else is wrong. 

I followed this https://docs.docker.com/registry/insecure/#using-self-signed-certificates and this https://docs.docker.com/registry/deploying/#running-a-domain-registry to setup a simple local Docker registry.

**Comment by uromahn** [comment]
I think I found the problem. The time in the pcf vm is UTC but instead of +8hrs it shows localtime -8 hrs (it says Mar 20, 14:47GMT instead of Mar 21, 06:47GMT which would be correct) which is strange. However, my certificate is not yet valid at the time inside the pcf VM.
I believe if I try it tomorrow, it should work.

**Comment by sclevine** [comment]
Thanks for the update. If that doesn't work, can your provide the result of running:
```
 cat /var/vcap/jobs/garden/bin/garden_ctl | grep docker
```
from inside the VM?

**Comment by uromahn** [comment]
Two points:

1. I figured out why the clock in the PCF DEV VM was off: I suspended and resumed it. It appears that the clock inside the VM does not get sync'ed with the host's clock, so I had to adjust it manually. Good to remember when doing a suspend/resume.

1. I tried it again and now I am getting the following inside `/var/vcap/sys/log/garden/garden.stdout.log`:
`"error":"Get https://host.pcfdev.io:5000/v2: x509: certificate signed by unknown authority"`
despite the fact that I started pcf dev with `cf dev start -r host.pcfdev.io:5000 -c 4 -m 6144`

Also a `cat /var/vcap/jobs/garden/bin/garden_ctl | grep docker` resulted in no line!

I have a feeling that the `-r` switch does not work properly with my setup.

**Comment by uromahn** [comment]
I took a look at assets/scripts/run and found the following:

```
if [[ -n "$registries" ]]; then
  perl -p -i -e "s/.*-insecureDockerRegistry=.*\n//i" /var/vcap/jobs/garden/bin/garden_ctl
  perl -p -i -e "s/.*-insecureDockerRegistry=.*\n//i" /var/vcap/jobs/stager/bin/stager_as_vcap
```
When I looked at `/var/vcap/jobs/stager/bin/stager_as_vcap` I found the line 
`-insecureDockerRegistry=host.pcfdev.io:5000 \` 
but it did not exist in `/var/vcap/jobs/garden/bin/garden_ctl`.

So it looks like a bug to me. :)

**Comment by sclevine** [comment]
Agreed! Thanks for the report. I can replicate this locally as well. Seems like a regression in a recent release.

As a workaround, you can manually add that line (see https://github.com/cloudfoundry/garden-runc-release/blob/develop/jobs/garden/templates/garden_ctl.erb#L284), and run `monit restart garden`.

(May be helpful to know, you can get root access on the VM with: `sudo su -`)

**Comment by uromahn** [comment]
Just some feedback on your last comment: workaround worked - thanks a ton!
Waiting for the fix in the next release now.

**Comment by mdelillo** [comment]
Hi @uromahn, thanks for reporting this. It has been fixed in the latest release (v0.26.0).
